### PR TITLE
feat: bulkdiscount

### PR DESCRIPTION
**Pull request by sahilk25**
https://app.asana.com/0/1192403191428092/1200413599428607/f

### CODE REVIEW

File: erpnext/selling/doctype/quoting_sheet/quoting_sheet.js [hunk 1]
```diff
@@ -50,6 +76,13 @@ frappe.ui.form.on("Quoting Sheet Item", {
 		if (row.qty && row.rate) {
 			let amount = row.qty * row.rate;
 			frappe.model.set_value(cdt, cdn, "amount", amount);
+			frappe.call({
+				method: "calculate_total_raw_material_cost",
+				doc: frm.doc,
+				callback: () => {
+					frm.refresh_field("rm_cost");
+				}
+			});
```
Comment (vishdha) [inline erpnext/selling/doctype/quoting_sheet/quoting_sheet.js]: line 64 to 70 and line 79 to 85 contain redundant code can you move it to a common function.

**Review by vishdha — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/quoting_sheet/quoting_sheet.py [hunk 2]
```diff
@@ -18,13 +20,42 @@ def calculate_totals(self):
...
+			if bom_items:
+				self.append("raw_material_items", {
+					"item_code": bom_items.get("item_code"),
+					"qty": bom_items.get("qty"),
+					"rate": bom_items.get("rate"),
+					"uom": bom_items.get("uom"),
+					"item_name": bom_items.get("item_name")
+				})
+		self.calculate_single_raw_material_cost()
+		self.calculate_total_raw_material_cost()
```
Comment (vishdha) [inline erpnext/selling/doctype/quoting_sheet/quoting_sheet.py]:
```suggestion
		for material in raw_materials:
			bom_item = frappe.db.get_value("BOM Item", material.name, ["item_code", "qty", "rate", "uom", "item_name"], as_dict = 1)
			if bom_item:
				self.append("raw_material_items", {
					"item_code": bom_item.get("item_code"),
					"qty": bom_item.get("qty"),
					"rate": bom_item.get("rate"),
					"uom": bom_item.get("uom"),
					"item_name": bom_item.get("item_name")
				})
		self.calculate_single_raw_material_cost()
		self.calculate_total_raw_material_cost() 
```

**Review by vishdha — CHANGES_REQUESTED** [review]

**Comment by rvpasari** [comment]
Don't see the doctype for bulk discounting here

File: erpnext/selling/doctype/quoting_sheet/quoting_sheet.js (hunk 1 above)
Comment (rvpasari) [reply erpnext/selling/doctype/quoting_sheet/quoting_sheet.js]: @sahilk25 could you look into this and update the PR?

File: erpnext/selling/doctype/quoting_sheet/quoting_sheet.py [hunk 3]
```diff
@@ -5,26 +5,87 @@
...
+		return self.rm_cost
+
+	def get_raw_materials(self):
+		"""
+			set raw materials items against to BOM Item given
+		"""
+		self.raw_material_items = []
+		raw_materials = frappe.get_all("BOM Item", filters={"parent": self.bom})
+		for material in raw_materials:
+			bom_item = frappe.db.get_value("BOM Item", material.name, ["item_code", "qty", "rate", "uom", "item_name"], as_dict = 1)
```
Comment (rvpasari) [inline erpnext/selling/doctype/quoting_sheet/quoting_sheet.py:54]:
```suggestion
			bom_item = frappe.db.get_value("BOM Item", material.name, fields=["item_code", "qty", "rate", "uom", "item_name"], as_dict = 1)
```

**Review by rvpasari — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/quoting_sheet/quoting_sheet.py (hunk 3 above)
Comment (sahilk25) [reply erpnext/selling/doctype/quoting_sheet/quoting_sheet.py:54]: @rvpasari fields NOT working here

File: erpnext/selling/doctype/quoting_sheet/quoting_sheet.js [hunk 4]
```diff
@@ -42,14 +71,26 @@ frappe.ui.form.on("Quoting Sheet Item", {
...
+			frm.trigger("calculate_total_raw_material_cost");
 		}
 	},
 
+	calculate_total_raw_material_cost: function (frm) {
+		frappe.call({
+			method: "calculate_total_raw_material_cost",
+			doc: frm.doc,
+			callback: () => {
+				frm.refresh_field("rm_cost");
```
Comment (hrwX) [inline erpnext/selling/doctype/quoting_sheet/quoting_sheet.js]: - Are we updating the value of rm_cost ?? If so I do not see a `set_value`

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/quoting_sheet/quoting_sheet.js (hunk 4 above)
Comment (sahilk25) [reply erpnext/selling/doctype/quoting_sheet/quoting_sheet.js]: @hrwX we are setting rm_cost value in python file, line no 44 and 45.
let me know if this not enough or something else.

**Comment by hrwX** [comment]
@sahilk25 Whats the status of this PR ?

**Comment by sahil28297** [comment]
@sahilk25 @rvpasari Any updates on this PR?

File: erpnext/selling/doctype/quoting_sheet/quoting_sheet.json [hunk 5]
```diff
@@ -155,12 +137,6 @@
    "options": "Quoting Sheet Item",
```
Comment (hrwX) [inline erpnext/selling/doctype/quoting_sheet/quoting_sheet.json:137]: - Is the `Customer Provided Item` position change pushed in this PR ?

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/selling/doctype/quoting_sheet/quoting_sheet.json (hunk 5 above)
Comment (sahilk25) [reply erpnext/selling/doctype/quoting_sheet/quoting_sheet.json:137]: Yes

File: erpnext/selling/doctype/quoting_sheet/quoting_sheet.json [hunk 6]
```diff
@@ -171,10 +147,46 @@
...
+   "options": "Bulk Discount Scheme"
+  },
+  {
+   "default": "0",
+   "fieldname": "profit_margin",
+   "fieldtype": "Percent",
+   "label": "Profit Margin %"
+  },
+  {
+   "fieldname": "quotation_to",
```
Comment (hrwX) [inline erpnext/selling/doctype/quoting_sheet/quoting_sheet.json]: - rename to quote_to since this is a Quote not a Quotation

**Review by hrwX — CHANGES_REQUESTED** [review]

**Review by hrwX — DISMISSED** [review]

**Review by sahil28297 — APPROVED** [review]

**Review by hrwX — APPROVED** [review]

**Pull request merged**